spacy/pipeline/_parser_internals/nonproj.pyx

```diff
@@ -49,39 +49,53 @@ def is_nonproj_arc(tokenid, heads):
     return _is_nonproj_arc(tokenid, c_heads)
 
 
-cdef bool _is_nonproj_arc(int tokenid, const vector[int]& heads) nogil:
+cdef int _is_nonproj_arc(int tokenid, const vector[int]& heads) nogil:
     # definition (e.g. Havelka 2007): an arc h -> d, h < d is non-projective
     # if there is a token k, h < k < d such that h is not
     # an ancestor of k. Same for h -> d, h > d
     head = heads[tokenid]
     if head == tokenid:  # root arcs cannot be non-projective
-        return False
+        return 0
     elif head < 0:  # unattached tokens cannot be non-projective
-        return False
+        return 0
     
     cdef int start, end
     if head < tokenid:
         start, end = (head+1, tokenid)
     else:
         start, end = (tokenid+1, head)
     for k in range(start, end):
-        if _has_head_as_ancestor(k, head, heads):
+        has_head = _has_head_as_ancestor(k, head, heads)
+        if has_head == 1:
             continue
-        else:  # head not in ancestors: d -> h is non-projective
-            return True
-    return False
+        elif has_head == 0:  # head not in ancestors: d -> h is non-projective
+            return 1
+        # -1 means "found cycle".
+        elif has_head == -1:
+            return -1
+    return 0
 
 
-cdef bool _has_head_as_ancestor(int tokenid, int head, const vector[int]& heads) nogil:
+cdef int _has_head_as_ancestor(int tokenid, int head, const vector[int]& heads) nogil:
     ancestor = tokenid
     cnt = 0
+    cdef unordered_set[int] seen_tokens
+    seen_tokens.insert(ancestor)
     while cnt < heads.size():
+        # Reached the head or a disconnected node
         if heads[ancestor] == head or heads[ancestor] < 0:
-            return True
+            return 1
+        # Reached the root
+        if heads[ancestor] == ancestor:
+            return 0
         ancestor = heads[ancestor]
+        # Found cycle
+        if seen_tokens.find(ancestor) != seen_tokens.const_end():
+            return -1
+        seen_tokens.insert(ancestor)
```

Comment:
One more optimization note. You could do this with a single lookup (code untested, but should give you an idea):

```suggestion
       result = seen_tokens.insert(ancestor)
        if not result.second:
            return -1
```

This works because `insert` returns a pair of an iterator pointing to the element corresponding to `ancestor` and a boolean indicating whether the value was inserted (which is `false` if the value was already in the set).

Reply:
Added a commit that does this.